UnidirectionalProxy.h

```diff
@@ -0,0 +1,188 @@
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef UNIDIRECTIONALPROXY_H_
+#define UNIDIRECTIONALPROXY_H_
+
+#include "rtos/Thread.h"
+#include "rtos/EventFlags.h"
+#include "util/CircularBuffer.h"
+#include "platform/CriticalSectionLock.h"
+
+/**
+ * Base class for an unidirectional proxy. It acquires and buffers data from
+ * one end point in interrupt context and in parallel transfer the data to the
+ * destination.
+ *
+ * Implementation needs to implement the functions:
+ *    * register_listener(): Set up the handler that will receive the data.
+ *    * transfer(const uint8_t* data, uint16_t length): transfer the data.
+ *
+ * The data handler in the implementation must also call the function
+ * acquire_data with a reader when data is ready to be read.
+ *
+ * @warning This class has been designed with performance in mind to avoid bytes
+ * loss over a transport non error resilient. Data are acquired and buffered in
+ * one context (preferably the interrupt context) while a real-time thread
+ * dequeue the data and transfer them. If your transport is resilient to error
+ * use another class that does everything in thread mode.
+ *
+ * @note CRTP is used instead of virtual dispatch to enhance performances.
+ *
+ * @warning Instances can be quite large; be careful of where they are
+ * allocated.
+ */
+template<class Impl>
+class UnidirectionalProxy : mbed::NonCopyable<UnidirectionalProxy<Impl> > {
+
+public:
+    /**
+     * Start the proxy thread that will handle the transfer of the data.
+     */
+    void start() {
+        impl()->register_listener();
+        _worker_thread.start(mbed::callback(this, &UnidirectionalProxy::run));
+    }
+
+protected:
+    UnidirectionalProxy() :
+        _worker_thread(osPriorityRealtime, sizeof(_thread_stack), _thread_stack),
+        _rx_buffer(),
+        _signal_channel()
+    { }
+
+    /**
+     * Acquire data from a reader.
+     *
+     * @param reader The instance that read the data. It must provide the following
+     * concepts:
+     *
+     *   - ready(): return true if the reader is ready to be read and false
+     *   if there is no data left to read.
+     *   - read(): return a chunk of data read; the type returned must be
+     *   acceptable by a CircularBuffer<uint8_t>. So it can be a byte or an
+     *   array view to a chunk of data read.
+     */
+    template<typename Reader>
+    void acquire_data(Reader& reader) {
```

Comment:
If we want efficient code generated then yes, this has to be templated; If we don't care about it then we can expect a "concrete" interface.

Note that template also relax the constraints on the return type of the read concept. As long as the return type it can be pushed into a CircularBuffer<uint8_t>, the implementation is free to return whatever it wants.